compiler/noirc_evaluator/src/ssa/plonky2_gen/console_asm_writer.rs

```diff
@@ -44,6 +44,58 @@ fn vecbooltarget2string(t: &Vec<BoolTarget>) -> String {
     result
 }
 
+fn targetslice2string(t: &[Target]) -> String {
+    let mut result = String::new();
+    write!(&mut result, "(");
+    let mut first = true;
+    for tt in t {
+        if first {
+            write!(&mut result, "{}", target2string(*tt));
+            first = false;
+        } else {
+            write!(&mut result, ",{}", target2string(*tt));
+        }
+    }
+    write!(&mut result, ")");
+    result
+
+}
```

Comment:
A Rust-ier way to achieve this is to implement the `Display` trait. Unfortunately, due to [E0210](https://doc.rust-lang.org/error_codes/E0210.html), there is the additional complication, that `&[Target]` needs to be wrapped in a new type (because we don't own `Target`). This [SO question](https://stackoverflow.com/questions/25413201/how-do-i-implement-a-trait-i-dont-own-for-a-type-i-dont-own) shows how. Also, I implement `Deref`, so we don't have to call `.0`. The following code achieves the same result as your code (I checked : ):

```
use std::fmt;
use std::ops::Deref;

struct TargetSlice<'a>(&'a [Target]);

impl<'a> Deref for TargetSlice<'a> {
    type Target = &'a [Target];

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl fmt::Display for TargetSlice<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "(")?;
        let mut first = true;
        for tt in **self {
            if first {
                write!(f, "{}", target2string(*tt))?;
                first = false;
            } else {
                write!(f, ",{}", target2string(*tt))?;
            }
        }
        write!(f, ")")
    }
}
```

The use site changes from this:

```
println!("register_public_inputs\t{}", targetslice2string(targets));
```

... to this:


```
println!("register_public_inputs\t{}", TargetSlice(targets));
```

---

I'm leaving the decision to you, to pick which one you like better. If you stick with your approach, which I could agree is simpler, I would at least ask that the functions names are named `targetSliceToString` instead of `targetslice2string`.

Reply:
I intentionally did not implement Display, because it suggests there's only one way to display the object. This means that we can't implement a different assembler writer that uses a different syntax to display the targets. Also, future versions of the PLONKY2 library might implement the Display trait for Target in their own way and this will change our output. And yet another problem is the fact that Rust requires the implementation of a trait for a type to be in the same crate as either the type definition, or the trait: https://doc.rust-lang.org/reference/items/implementations.html#trait-implementation-coherence This means the Display implementation needs to be added to the same crate as the PLONKY2 library. I agree about fixing the naming convention for targetslice2string, etc.

Reply:
Hmm, speaking of wrapping &[Target] with a new type, we can also wrap Target with a new type, that is specific to the asm writer and implement Display for it.

Reply:
Ok, I wrapped all objects with new types and implemented Display for them in: https://github.com/blocksense-network/noir/pull/19/commits/f5ad284880f501f96f1263fbfd1a1decc82b3aa1

Reply:
Thanks, I love it!